net/tcp/tcp_send_buffered.c

```diff
@@ -528,7 +527,7 @@ static uint16_t psock_send_eventhandler(FAR struct net_driver_s *dev,
                 {
                   /* Do fast retransmit */
 
-                  rexmit = true;
+                  flags |= TCP_REXMIT;
```

Comment:
Concerning TCP protocol I am relying on RFCs only. Could you provide a link to RFC for retransmitting all the segments?

Reply:
rfc5681 does not explicitly state that only first unacked segments are retransmitted. I think it is reasonable to retransmit only the earliest segment under fast retransmission. However, if the retransmission is triggered by tcp timer, I think the original logic can be maintained.
https://github.com/apache/incubator-nuttx/blob/master/net/tcp/tcp_timer.c#L352

Reply:
E.g. RFC 5681 3.2.3 that you have referred says "The lost segment starting at SND.UNA MUST be retransmitted ...".
The word "segment" is not plural. That means only one segment.

RFC 6298 5.4 (that I have referred in the PR description) says:
> When the retransmission timer expires, do the following:
>
>   (5.4) Retransmit the earliest segment that has not been acknowledged
         by the TCP receiver.

All is written explicitly.

Reply:
But only retransmitting the earliest unacked packet is not fully applicable to the current stack implementation, especially when several consecutive segments are dropped.
Do you have a real device with a wireless module? I think you can test your commit with different decays

Reply:
> But only retransmitting the earliest unacked packet is not fully applicable to the current stack implementation, especially when several consecutive segments are dropped.

Why?

> Do you have a real device with a wireless module? I think you can test your commit with different decays

Yes, I do. I have tested tcp_send_buffered (this PR) and tcp_send_unbuffered (#4659 where I have fixed the same issue the same way) on real hardware. It worked just fine concerning TCP retransmissions. Before the applied patches streaming the data over TCP had serious lags (about 1 second) in case of TCP retransmissions, especially at high network traffic.

Reply:
my suggestion is:

1. For the retransmission triggered by RTO, maintain the logic of all retransmissions:
https://github.com/apache/incubator-nuttx/blob/master/net/tcp/tcp_timer.c#L352

This is a safeguard to guarantee that all data will definitely be retransmitted.

2. If retransmission triggered by fast retransmission, follow RFC 6298 (5.4) only one of the earliest not acknowledged segment should be retransmitted

Improvement of this PR (#5589) , your test will most likely be in this case, if there is a problem with fast retransmission, then RTO will guarantee that all segments are retransmitted.